pkg/prometheus/operator.go

```diff
@@ -1299,7 +1300,7 @@ func (c *Operator) sync(ctx context.Context, key string) error {
 
 		if ok && sErr.ErrStatus.Code == 422 && sErr.ErrStatus.Reason == metav1.StatusReasonInvalid {
 			c.metrics.StsDeleteCreateCounter().Inc()
-			level.Info(c.logger).Log("msg", "resolving illegal update of Prometheus StatefulSet", "details", sErr.ErrStatus.Details)
+			level.Info(c.logger).Log("msg", "recreating Prometheus StatefulSet because the update operation wasn't possible", "reason", sErr.ErrStatus.Details.Causes[0].Message)
```

Comment:
while you're at it, it might be good to add the `statefulset/shard` keys to all log messages as in here:
 https://github.com/prometheus-operator/prometheus-operator/blob/d26fd2d25e85b16b6ff944daf979de2b6d9e09ac/pkg/prometheus/operator.go#L1256

you can assign `logger := log.With(c.logger, "statefulset", ssetName, "shard", fmt.Sprintf("%d", shard))` at the beginning of the for loop and replace `c.logger` by `logger`.